### PR TITLE
Stats UI and 7 week lookback

### PR DESCRIPTION
**Pull request by diegocepedaw**
Add a a frontend for all the stats, create high priority incident stats endpoint, return all stats with 7 week lookback

### CODE REVIEW

File: src/iris/api.py [hunk 1]
```diff
@@ -4243,18 +4243,68 @@ def on_get(self, req, resp):
...
+
+    def on_get(self, req, resp):
+
+        connection = db.engine.raw_connection()
+        cursor = connection.cursor()
+        stats = {}
+        if self.real_time:
+            stats = app_stats.calculate_high_priority_incidents(connection, cursor)
+        else:
+            cursor.execute('SELECT timestamp, name, value FROM application_stats JOIN application on application_stats.application_id = application.id WHERE statistic = "high_priority_incidents_last_week" ORDER BY value DESC')
```
Comment (jrgp) [inline src/iris/api.py]: can this query be broken into multiple lines so it becomes easier to read?

File: src/iris/api.py [hunk 2]
```diff
@@ -4243,18 +4243,68 @@ def on_get(self, req, resp):
...
+            if cursor.rowcount == 0:
+                logger.exception('Error retrieving hpi stats from db')
+                raise HTTPInternalServerError('Error retrieving stats from db')
```
Comment (jrgp) [inline src/iris/api.py]: If this raises, the connection and cursor close lines will never get called, leading to us potentially leaking connetions.

Also, why is a rowcount of 0 worthy of a 5xx? 5xx usually refers to things actually being broken / database exceptions being raised

File: src/iris/api.py (hunk 1 above)
Comment (diegocepedaw) [reply src/iris/api.py]: I'll split it up

File: src/iris/api.py (hunk 2 above)
Comment (diegocepedaw) [reply src/iris/api.py]: The reason it's raising a 5xx is because unless no apps exist in the database the rowcount should never be 0 here. If the rowcount is 0 it means that the app_stats script failed and the stats tables are in a bad state.

File: src/iris/app_stats.py [hunk 3]
```diff
@@ -84,116 +98,180 @@ def calculate_app_stats(app, connection, cursor, fields_filter=None):
...
+def calculate_high_priority_incidents(connection, cursor):
+
+    high_priority_query = '''SELECT application.name, COUNT(DISTINCT incident.application_id, incident.id) AS incident_count FROM incident JOIN application ON application.id = incident.application_id JOIN message ON message.incident_id = incident.id WHERE incident.created > %(date_variable)s - INTERVAL 1 WEEK AND incident.created < %(date_variable)s AND priority_id IN ( SELECT id FROM priority WHERE name IN ('high', 'urgent') ) GROUP BY incident.application_id ORDER BY incident_count DESC'''
+
+    stats = {}
+
+    for delta in range(0, 7):
+        now = datetime.utcnow() - timedelta(weeks=delta)
+        formatted_date = now.strftime('%Y-%m-%d %H:%M:%S')
+        stats[formatted_date] = []
```
Comment (dwang159) [inline src/iris/app_stats.py]: When we're passing datetimes around, we should always use Unix timestamps to avoid timezone shenanigans. Converting to a string may/may not be necessary when talking to the DB, you'll need to double check that. For the stats variable, we should keep it in Unix time, then convert to pretty print in the frontend if needed.

File: src/iris/app_stats.py [hunk 4]
```diff
@@ -84,116 +98,180 @@ def calculate_app_stats(app, connection, cursor, fields_filter=None):
...
+            if result:
+                result = result[-1]
+            if result is None:
+                result = -1
+
+            logger.info('Stats query %s took %s seconds', key, round(time.time() - start, 2))
+
+            # {statistic : {timestamp: value, timestamp: value}}
+            if stats.get(key):
+                stats[key][formatted_date] = result
```
Comment (dwang159) [inline src/iris/app_stats.py]: Same here for the date stuff

File: src/iris/api.py [hunk 5]
```diff
@@ -4265,26 +4315,41 @@ def __init__(self, config):
         self.real_time = cfg.get('real_time', True)
 
     def on_get(self, req, resp, app_name):
-        app = cache.applications.get(app_name)
```
[the remark this answers is not in the document]
Comment (diegocepedaw) [reply src/iris/api.py]: The stats endpoint is such a low qps endpoint that I didn't see much justification for using the application cache here. Plus the cache is currently case sensitive which is not ideal for checking the application name in the url aginst.

File: src/iris/api.py [hunk 6]
```diff
@@ -4243,18 +4243,68 @@ def on_get(self, req, resp):
...
+                    stats[row[0]] = {}
+                    stats[row[0]][row[2]] = row[1]
 
         cursor.close()
         connection.close()
         resp.status = HTTP_200
         resp.body = ujson.dumps(stats, sort_keys=True)
 
 
+class HighPriorityIncidents(object):
```
[the remark this answers is not in the document]
Comment (diegocepedaw) [reply src/iris/api.py]: I considered handling as /stats/{statistic} but since high priority incidents per week was the only statistic highlighted as being of particular importance for weekly reporting I decided to avoid naming conflicts and just make it its own endpoint.

File: src/iris/app_stats.py [hunk 7]
```diff
@@ -28,44 +29,57 @@ def calculate_app_stats(app, connection, cursor, fields_filter=None):
...
+        query_data = {'application_id': app['id'], 'date_variable': formatted_date}
+
+        for key in fields:
+            start = time.time()
+            cursor.execute(queries[key], query_data)
+            result = cursor.fetchone()
+            if result:
+                result = result[-1]
+            if result is None:
+                result = -1
```
[the remark this answers is not in the document]
Comment (diegocepedaw) [reply src/iris/app_stats.py]: The schema originally defined the value field as NOT NULL but considering how I've changed the behavior it probably makes more sense to change that to allow nulls and change the -1s to None.

File: src/iris/app_stats.py [hunk 8]
```diff
@@ -84,116 +68,226 @@ def calculate_app_stats(app, connection, cursor, fields_filter=None):
...
+                            GROUP BY `mode_id`) `counts`
+                        JOIN `mode` ON `mode`.`id` = `counts`.`mode_id`''',
+                       query_data)
+
+        for row in cursor:
+            key = 'total_%s_sent_last_week' % row[0]
+            result = row[1]
+
+            if stats.get(key):
+                stats[key].appennd({unix_date: result})
```
Comment (dwang159) [inline src/iris/app_stats.py]: `appennd` is a typo here and on line 169. We should fix this typo, make sure the rest of the script has been tested thoroughly, and try to patch the gaps in our e2e testing to make sure we don't miss things like this.

File: src/iris/api.py [hunk 9]
```diff
@@ -4253,12 +4253,12 @@ def on_get(self, req, resp):
             stats = {}
 
             for row in cursor:
-                # {statistic : {timestamp: value, timestamp: value}}
+                # format: {statistic : [{timestamp: value}, {timestamp: value}]}
```
Comment (dwang159) [inline src/iris/api.py]: Why the change in format? I don't particularly care for one over the other, but curious why it changed.

File: src/iris/bin/app_stats.py [hunk 10]
```diff
@@ -44,22 +45,30 @@ def set_global_stats(stats, connection, cursor):
...
-            cursor.execute('''INSERT INTO `global_stats` (`statistic`, `value`, `timestamp`)
-                            VALUES (%s, %s, %s)
-                            ON DUPLICATE KEY UPDATE `value`= %s, `timestamp` = %s''',
-                           (stat, val, timestamp, val, timestamp))
+    # format: {statistic : [{timestamp: value}, {timestamp: value}]}
+    for stat, entry_list in stats.items():
+        for stat_entry in entry_list:
+            for timestamp, val in stat_entry.items():
+                # format timestamp into datetime to store in mysql
+                timestamp = datetime.utcfromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')
```
Comment (dwang159) [inline src/iris/bin/app_stats.py]: For future reference, `FROM_UNIXTIME()` will do this in SQL, so you don't have to mess with datetime and timestring formats.

File: test/e2etest.py [hunk 11]
```diff
@@ -2160,38 +2160,14 @@ def test_healthcheck():
 def test_stats():
     re = requests.get(base_url + 'stats')
     assert re.status_code == 200
-    data = re.json()
```
Comment (dwang159) [inline test/e2etest.py]: Speaking of testing, are there any validation checks we can do on the data outside of just checking for HTTP 200? Something like what's deleted here, to make sure we're returning data that looks like it's correct. Doesn't need to be as extensive; spot checking is probably fine.

File: src/iris/api.py (hunk 9 above)
Comment (diegocepedaw) [reply src/iris/api.py]: I changed it to use lists so it would preserve sorted order when rendered by the browser.
